crates/time_format/src/time_format.rs

```diff
@@ -0,0 +1,243 @@
+        None => format!("{:02}:{:02}", hour, timestamp_local_minute),
+    };
+
+    let reference_local = reference.to_offset(timezone);
+    let reference_local_date = reference_local.date();
+    let timestamp_local_date = timestamp_local.date();
+
+    if timestamp_local_date == reference_local_date {
+        return formatted_time;
+    }
+
+    if reference_local_date.previous_day() == Some(timestamp_local_date) {
+        return format!("yesterday at {}", formatted_time);
+    }
+
+    match meridiem {
+        Some(_) => format!(
+            "{:02}/{:02}/{}",
+            timestamp_local_date.month() as u32,
+            timestamp_local_date.day(),
+            timestamp_local_date.year()
+        ),
+        None => format!(
+            "{:02}/{:02}/{}",
+            timestamp_local_date.day(),
+            timestamp_local_date.month() as u32,
+            timestamp_local_date.year()
+        ),
+    }
+}
+
+/// Returns true if the users configuration is set to 12-hour clock,
+/// or if the locale is recognized as a 12-hour clock locale.
+pub fn is_12_hour_clock() -> bool {
+    IS_12_HOUR_CLOCK_USER_PREFERENCE
+        .as_ref()
+        .map(|c| *c)
+        .unwrap_or_else(|_| is_12_hour_clock_by_locale(CURRENT_LOCALE.as_str()))
+}
+
+/// Returns true if the locale is recognized as a 12-hour clock locale.
+pub fn is_12_hour_clock_by_locale(locale: &str) -> bool {
+    [
+        "es-MX", "es-CO", "es-SV", "es-NI",
+        "es-HN", // Mexico, Colombia, El Salvador, Nicaragua, Honduras
+        "en-US", "en-CA", "en-AU", "en-NZ", // U.S, Canada, Australia, New Zealand
+        "ar-SA", "ar-EG", "ar-JO", // Saudi Arabia, Egypt, Jordan
+        "en-IN", "hi-IN", // India, Hindu
+        "en-PK", "ur-PK", // Pakistan, Urdu
+        "en-PH", "fil-PH", // Philippines, Filipino
+        "bn-BD", "ccp-BD", // Bangladesh, Chakma
+        "en-IE", "ga-IE", // Ireland, Irish
+        "en-MY", "ms-MY", // Malaysia, Malay
+    ]
+    .contains(&locale)
+}
+
+lazy_static! {
+    static ref CURRENT_LOCALE: String =
+        sys_locale::get_locale().unwrap_or_else(|| String::from("en-US"));
+    static ref IS_12_HOUR_CLOCK_USER_PREFERENCE: Result<bool> =
+        read_is_12_hour_clock_from_user_preferences();
+}
+
+fn read_is_12_hour_clock_from_user_preferences() -> Result<bool> {
+    #[cfg(target_os = "macos")]
+    {
+        let mut file_path = util::paths::HOME.clone();
+        file_path.push("Library");
+        file_path.push("Preferences");
+        file_path.push(".GlobalPreferences.plist");
+
+        let value: plist::Value = plist::from_file(file_path.as_path())?;
+        if let Some(root_dict) = value.as_dictionary() {
+            if let Some(plist::Value::Boolean(force_12_hour_format)) =
+                root_dict.get("AppleICUForce12HourTime")
+            {
+                return Ok(*force_12_hour_format);
+            }
+        }
+        Ok(false)
```

Comment:
@bennetbo This doesn't seem to work for me: I have a US locale and have disabled the 24-hour override setting.

Rather than trying to figure this out for ourselves, I suggest we use CoreFoundation's builtin time formatting code.

We could either call into it directly, or use it to calculate the current format for us and then continue to format ourselves.

We need to add a dependency on `core-foundation-sys` and then something like the following code works:

```
        let str = unsafe {
            let locale = CFLocaleCopyCurrent();
            let formatter = CFDateFormatterCreate(
                kCFAllocatorDefault,
                locale,
                kCFDateFormatterNoStyle,
                kCFDateFormatterShortStyle,
            );
            let fmt = CFDateFormatterGetFormat(formatter);

            let str = CFType::wrap_under_get_rule(fmt.as_void_ptr())
                .downcast::<CFString>()
                .ok_or_else(|| anyhow::anyhow!(dbg!("not a string")))?;

            // CFRelease(locale); 
            // CFRelease(formatter);

            str.to_string()
        };
        dbg!(str); // "h:mm a" when I have the 24-hour thing disabled, or "HH:mm" when it is enabled.
```

It might be simpler to go one step further and call `CFDateFormatterCreateStringWithAbsoluteTime` directly on mac to generate the timestamps; but it will depend on how much code is needed to convert between rust dates and cocoa dates (and in the opposite direction with strings).

Reply:
I installed `en_US` on my system and can confirm that this does not seem to work. Interestingly neither AppleICUForce12HourTime or AppleICUForce24HourTime seem to be present when using `en_US` and 12 hour or 24 hour time formatting. I will look into the CFDateFormatter approach, I believe this will also handle more advanced cases, f.e. you can customize the date formatting in macOS converting `01/01/20` to `01.01.20` which I believe is more common in European locales.